cmd/ncproxy/run.go

```diff
@@ -0,0 +1,200 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"syscall"
+	"time"
+
+	"github.com/Microsoft/go-winio/pkg/etwlogrus"
+	"github.com/Microsoft/hcsshim/cmd/ncproxy/nodenetsvc"
+	"github.com/Microsoft/hcsshim/internal/computeagent"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/oc"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"go.opencensus.io/plugin/ocgrpc"
+	"go.opencensus.io/trace"
+	"google.golang.org/grpc"
+)
+
+type nodeNetSvcConn struct {
+	client   nodenetsvc.NodeNetworkServiceClient
+	addr     string
+	grpcConn *grpc.ClientConn
+}
+
+var (
+	// Global mapping of network namespace ID to shim compute agent ttrpc service.
+	containerIDToShim = make(map[string]computeagent.ComputeAgentService)
+	// Global object representing the connection to the node network service that
+	// ncproxy will be talking to.
+	nodeNetSvcClient *nodeNetSvcConn
+)
+
+var (
+	configPath    = flag.String("config", "", "Path to JSON configuration file.")
+	logDir        = flag.String("log-directory", "", "Directory to write ncproxy logs to. This is just panic logs.")
+	registerSvc   = flag.Bool("register-service", false, "Register ncproxy as a Windows service.")
+	unregisterSvc = flag.Bool("unregister-service", false, "Unregister ncproxy as a Windows service.")
+	runSvc        = flag.Bool("run-service", false, "Run ncproxy as a Windows service.")
+)
+
+// Run ncproxy
+func run() error {
+	logrus.SetOutput(ioutil.Discard)
+	flag.Parse()
+
+	// For both unregistering and registering the service we need to exit out (even on success). -register-service will register
+	// ncproxy's commandline to launch with the -run-service flag set.
+	if *unregisterSvc {
+		if *registerSvc {
+			return errors.New("-register-service and -unregister-service cannot be used together")
+		}
+		return unregisterService()
+	}
+
+	if *registerSvc {
+		return registerService()
+	}
+
+	// If no logging directory passed in use where ncproxy is located.
+	if *logDir == "" {
+		binLocation, err := os.Executable()
+		if err != nil {
+			return err
+		}
+		*logDir = filepath.Dir(binLocation)
+	} else {
+		// If a log dir was provided, make sure it exists.
+		if _, err := os.Stat(*logDir); os.IsNotExist(err) {
```

Comment:
should we return errors that aren't os.IsNotExist?

Reply:
I guess we should just check err != nil as mkdirall will just return nil if the directory already exists.